File: api-tester/src/main/java/com/revenuecat/apitester/java/StoreProductAPI.java
package com.revenuecat.apitester.java;

import com.android.billingclient.api.ProductDetails;
import com.revenuecat.purchases.ProductType;
import com.revenuecat.purchases.amazon.AmazonStoreProduct;
import com.revenuecat.purchases.amazon.AmazonStoreProductKt;
import com.revenuecat.purchases.models.GoogleStoreProduct;
import com.revenuecat.purchases.models.GoogleStoreProductKt;
import com.revenuecat.purchases.models.Period;
import com.revenuecat.purchases.models.Price;
import com.revenuecat.purchases.models.StoreProduct;
import com.revenuecat.purchases.models.SubscriptionOption;
import com.revenuecat.purchases.models.SubscriptionOptions;

import org.json.JSONObject;

import java.util.Locale;

@SuppressWarnings({"unused"})
final class StoreProductAPI {
    static void check(final StoreProduct product) {
        final Locale locale = Locale.getDefault();
        final String productId = product.getId();
        final String sku = product.getSku();
        final ProductType type = product.getType();
        final Price price = product.getPrice();
        final String title = product.getTitle();
        final String description = product.getDescription();
        final Period period = product.getPeriod();
        final String formattedPricePerMonth = product.formattedPricePerMonth(locale);
        final Price pricePerWeek = product.pricePerWeek(locale);
        final Price pricePerMonth = product.pricePerMonth(locale);
        final Price pricePerYear = product.pricePerYear(locale);

        SubscriptionOptions subscriptionOptions = product.getSubscriptionOptions();
        SubscriptionOption defaultOption = product.getDefaultOption();
        String presentedOfferingIdentifier = product.getPresentedOfferingIdentifier();

        GoogleStoreProduct underlyingGoogleProduct = GoogleStoreProductKt.getGoogleProduct(product);
        AmazonStoreProduct underlyingAmazonProduct = AmazonStoreProductKt.getAmazonProduct(product);
    }

    static void check(final ProductType type) {
        switch (type) {
            case SUBS:
            case INAPP:
            case UNKNOWN:
        }
    }

    static void checkGoogleStoreProduct(GoogleStoreProduct googleStoreProduct) {
        check(googleStoreProduct);
        GoogleStoreProduct constructedGoogleStoreProduct = new GoogleStoreProduct(
                googleStoreProduct.getId(),
                googleStoreProduct.getBasePlanId(),
                googleStoreProduct.getType(),
                googleStoreProduct.getPrice(),
                googleStoreProduct.getTitle(),
                googleStoreProduct.getDescription(),
                googleStoreProduct.getPeriod(),
                googleStoreProduct.getSubscriptionOptions(),
                googleStoreProduct.getDefaultOption(),
                googleStoreProduct.getProductDetails(),
                googleStoreProduct.getPresentedOfferingIdentifier()
        );

        GoogleStoreProduct constructedGoogleStoreProductWithNullableFieldsNull = new GoogleStoreProduct(
                googleStoreProduct.getId(),
                null,
                googleStoreProduct.getType(),
                googleStoreProduct.getPrice(),
                googleStoreProduct.getTitle(),
                googleStoreProduct.getDescription(),
                null,
                null,
                null,
                googleStoreProduct.getProductDetails(),
                null
        );

        String productId = constructedGoogleStoreProduct.getProductId();
        String basePlanId = constructedGoogleStoreProduct.getBasePlanId();
        ProductDetails productDetails = googleStoreProduct.getProductDetails();
    }

    static void checkAmazonStoreProduct(AmazonStoreProduct amazonStoreProduct) {
        check(amazonStoreProduct);
        AmazonStoreProduct constructedAmazonStoreProduct = new AmazonStoreProduct(
                amazonStoreProduct.getId(),
                amazonStoreProduct.getType(),
                amazonStoreProduct.getTitle(),
                amazonStoreProduct.getDescription(),
                amazonStoreProduct.getPeriod(),
                amazonStoreProduct.getPrice(),
                amazonStoreProduct.getSubscriptionOptions(),
                amazonStoreProduct.getDefaultOption(),
                amazonStoreProduct.getIconUrl(),
                amazonStoreProduct.getFreeTrialPeriod(),
                amazonStoreProduct.getOriginalProductJSON(),
                amazonStoreProduct.getPresentedOfferingIdentifier()
        );

        AmazonStoreProduct constructedAmazonStoreProductWithNullOfferingId = new AmazonStoreProduct(
                amazonStoreProduct.getId(),
                amazonStoreProduct.getType(),
                amazonStoreProduct.getTitle(),
                amazonStoreProduct.getDescription(),
                null,
                amazonStoreProduct.getPrice(),
                null,
                null,
                amazonStoreProduct.getIconUrl(),
                null,
                amazonStoreProduct.getOriginalProductJSON(),
                null
        );
        String iconUrl = amazonStoreProduct.getIconUrl();
        Period freeTrialPeriod = amazonStoreProduct.getFreeTrialPeriod();
        JSONObject originalProductJson = amazonStoreProduct.getOriginalProductJSON();
    }
}

File: api-tester/src/main/java/com/revenuecat/apitester/kotlin/StoreProductAPI.kt
package com.revenuecat.apitester.kotlin

import com.android.billingclient.api.ProductDetails
import com.revenuecat.purchases.PresentedOfferingContext
import com.revenuecat.purchases.ProductType
import com.revenuecat.purchases.amazon.AmazonStoreProduct
import com.revenuecat.purchases.amazon.amazonProduct
import com.revenuecat.purchases.models.GoogleStoreProduct
import com.revenuecat.purchases.models.Period
import com.revenuecat.purchases.models.Price
import com.revenuecat.purchases.models.StoreProduct
import com.revenuecat.purchases.models.SubscriptionOption
import com.revenuecat.purchases.models.SubscriptionOptions
import com.revenuecat.purchases.models.googleProduct
import org.json.JSONObject
import java.util.Locale

@Suppress("unused", "UNUSED_VARIABLE", "LongMethod", "DEPRECATION")
private class StoreProductAPI {
    fun check(product: StoreProduct) {
        val locale = Locale.getDefault()
        with(product) {
            val storeProductId: String = id
            val sku: String = sku
            val type: ProductType = type
            val price: Price = price
            val formattedPricePerMonth: String? = formattedPricePerMonth(locale)
            val formattedPricePerMonthNoLocale: String? = formattedPricePerMonth()
            val pricePerWeek: Price? = pricePerWeek(locale)
            val pricePerMonth: Price? = pricePerMonth(locale)
            val pricePerYear: Price? = pricePerYear(locale)
            val pricePerWeekNoLocale: Price? = pricePerYear()
            val pricePerMonthNoLocale: Price? = pricePerMonth()
            val pricePerYearNoLocale: Price? = pricePerYear()
            val name: String = name
            val title: String = title
            val description: String = description
            val period: Period? = period
            val subscriptionOptions: SubscriptionOptions? = subscriptionOptions
            val defaultOption: SubscriptionOption? = defaultOption
            val presentedOfferingIdentifier: String? = presentedOfferingIdentifier
            val presentedOfferingContext: PresentedOfferingContext? = presentedOfferingContext
            val underlyingGoogleProduct: GoogleStoreProduct? = googleProduct
            val underlyingAmazonProduct: AmazonStoreProduct? = amazonProduct
            val copiedProductWithOfferingId: StoreProduct = copyWithOfferingId("offeringId")
            val copiedProduct: StoreProduct = copyWithPresentedOfferingContext(null)
            val copiedProduct2: StoreProduct = copyWithPresentedOfferingContext(
                PresentedOfferingContext(
                    offeringIdentifier = "offeringId",
                    placementIdentifier = "placementId",
                    targetingRevision = 1,
                    targetingRuleId = "ruleId",
                ),
            )
        }
    }

    fun check(type: ProductType) {
        when (type) {
            ProductType.SUBS,
            ProductType.INAPP,
            ProductType.UNKNOWN,
            -> {}
        }.exhaustive
    }

    fun checkGoogleStoreProduct(googleStoreProduct: GoogleStoreProduct) {
        check(googleStoreProduct)
        val constructedGoogleStoreProduct = GoogleStoreProduct(
            googleStoreProduct.id,
            null,
            googleStoreProduct.type,
            googleStoreProduct.price,
            googleStoreProduct.title,
            googleStoreProduct.description,
            googleStoreProduct.period,
            googleStoreProduct.subscriptionOptions,
            googleStoreProduct.defaultOption,
            googleStoreProduct.productDetails,
        )

        val constructedGoogleStoreProductWithNullOfferingId = GoogleStoreProduct(
            googleStoreProduct.id,
            null,
            googleStoreProduct.type,
            googleStoreProduct.price,
            googleStoreProduct.name,
            googleStoreProduct.title,
            googleStoreProduct.description,
            googleStoreProduct.period,
            googleStoreProduct.subscriptionOptions,
            googleStoreProduct.defaultOption,
            googleStoreProduct.productDetails,
            null,
        )

        val constructedGoogleStoreProductWithOfferingId = GoogleStoreProduct(
            googleStoreProduct.id,
            null,
            googleStoreProduct.type,
            googleStoreProduct.price,
            googleStoreProduct.title,
            googleStoreProduct.description,
            googleStoreProduct.period,
            googleStoreProduct.subscriptionOptions,
            googleStoreProduct.defaultOption,
            googleStoreProduct.productDetails,
            "offeringId",
        )

        val constructedGoogleStoreProductWithName = GoogleStoreProduct(
            googleStoreProduct.id,
            null,
            googleStoreProduct.type,
            googleStoreProduct.price,
            googleStoreProduct.name,
            googleStoreProduct.title,
            googleStoreProduct.description,
            googleStoreProduct.period,
            googleStoreProduct.subscriptionOptions,
            googleStoreProduct.defaultOption,
            googleStoreProduct.productDetails,
        )

        val constructedGoogleStoreProductWithNameButNoOfferingId = GoogleStoreProduct(
            googleStoreProduct.id,
            null,
            googleStoreProduct.type,
            googleStoreProduct.price,
            googleStoreProduct.name,
            googleStoreProduct.title,
            googleStoreProduct.description,
            googleStoreProduct.period,
            googleStoreProduct.subscriptionOptions,
            googleStoreProduct.defaultOption,
            googleStoreProduct.productDetails,
        )

        val productId: String = constructedGoogleStoreProduct.productId
        val basePlanId: String? = constructedGoogleStoreProduct.basePlanId
        val productDetails: ProductDetails = googleStoreProduct.productDetails
    }

    fun checkAmazonStoreProduct(amazonStoreProduct: AmazonStoreProduct) {
        check(amazonStoreProduct)
        val constructedAmazonStoreProduct = AmazonStoreProduct(
            amazonStoreProduct.id,
            amazonStoreProduct.type,
            amazonStoreProduct.title,
            amazonStoreProduct.description,
            amazonStoreProduct.period,
            amazonStoreProduct.price,
            amazonStoreProduct.subscriptionOptions,
            amazonStoreProduct.defaultOption,
            amazonStoreProduct.iconUrl,
            amazonStoreProduct.freeTrialPeriod,
            amazonStoreProduct.originalProductJSON,
        )

        val constructedAmazonStoreProductWithContext = AmazonStoreProduct(
            amazonStoreProduct.id,
            amazonStoreProduct.type,
            amazonStoreProduct.name,
            amazonStoreProduct.title,
            amazonStoreProduct.description,
            amazonStoreProduct.period,
            amazonStoreProduct.price,
            amazonStoreProduct.subscriptionOptions,
            amazonStoreProduct.defaultOption,
            amazonStoreProduct.iconUrl,
            amazonStoreProduct.freeTrialPeriod,
            amazonStoreProduct.originalProductJSON,
            PresentedOfferingContext("offeringId"),
        )

        val constructedAmazonStoreProductWithNullContext = AmazonStoreProduct(
            amazonStoreProduct.id,
            amazonStoreProduct.type,
            amazonStoreProduct.name,
            amazonStoreProduct.title,
            amazonStoreProduct.description,
            amazonStoreProduct.period,
            amazonStoreProduct.price,
            amazonStoreProduct.subscriptionOptions,
            amazonStoreProduct.defaultOption,
            amazonStoreProduct.iconUrl,
            amazonStoreProduct.freeTrialPeriod,
            amazonStoreProduct.originalProductJSON,
            null,
        )

        val constructedAmazonStoreProductWithOfferingId = AmazonStoreProduct(
            amazonStoreProduct.id,
            amazonStoreProduct.type,
            amazonStoreProduct.title,
            amazonStoreProduct.description,
            amazonStoreProduct.period,
            amazonStoreProduct.price,
            amazonStoreProduct.subscriptionOptions,
            amazonStoreProduct.defaultOption,
            amazonStoreProduct.iconUrl,
            amazonStoreProduct.freeTrialPeriod,
            amazonStoreProduct.originalProductJSON,
            amazonStoreProduct.presentedOfferingIdentifier,
        )

        val constructedAmazonStoreProductWithName = AmazonStoreProduct(
            amazonStoreProduct.id,
            amazonStoreProduct.type,
            amazonStoreProduct.name,
            amazonStoreProduct.title,
            amazonStoreProduct.description,
            amazonStoreProduct.period,
            amazonStoreProduct.price,
            amazonStoreProduct.subscriptionOptions,
            amazonStoreProduct.defaultOption,
            amazonStoreProduct.iconUrl,
            amazonStoreProduct.freeTrialPeriod,
            amazonStoreProduct.originalProductJSON,
        )

        val constructedAmazonStoreProductWithNameAndOfferingId = AmazonStoreProduct(
            amazonStoreProduct.id,
            amazonStoreProduct.type,
            amazonStoreProduct.name,
            amazonStoreProduct.title,
            amazonStoreProduct.description,
            amazonStoreProduct.period,
            amazonStoreProduct.price,
            amazonStoreProduct.subscriptionOptions,
            amazonStoreProduct.defaultOption,
            amazonStoreProduct.iconUrl,
            amazonStoreProduct.freeTrialPeriod,
            amazonStoreProduct.originalProductJSON,
            amazonStoreProduct.presentedOfferingContext,
        )
        val iconUrl: String = amazonStoreProduct.iconUrl
        val freeTrialPeriod: Period? = amazonStoreProduct.freeTrialPeriod
        val originalProductJson: JSONObject = amazonStoreProduct.originalProductJSON
    }
}
